File: efdi-moon-pod-main/clients/examples/modern/typescript/publish.ts
// publish.ts — send data to the goat fabric (modern TypeScript).
//
//   npm install && npm run publish              # one JSON sample
//   npm run publish -- 50 0.2                    # 50 samples at 200ms
//
// Publishes JSON under <namespace>/sensors/temp via the pod's remote-api plugin (WebSocket).
// Real payloads can be anything (bytes, protobuf, CBOR); JSON here for legibility.
//
// IMPORTANT: zenoh-ts connects to the zenoh-plugin-remote-api over a WebSocket, NOT directly to
// the router over mTLS. See ./README.md and ../../../connect/typescript/goat_connect.ts.

import { Encoding } from "@eclipse-zenoh/zenoh-ts";
import { session, key } from "../../../connect/typescript/goat_connect.js";

async function main(): Promise<void> {
  const n = process.argv[2] ? parseInt(process.argv[2], 10) : 1;
  const interval = process.argv[3] ? parseFloat(process.argv[3]) : 1.0;
  const k = key("sensors/temp");

  const s = await session();
  try {
    const pub = await s.declarePublisher(k, { encoding: Encoding.APPLICATION_JSON });
    for (let i = 0; i < n; i++) {
      const payload = JSON.stringify({ ts: Date.now(), seq: i, temp_c: 21.5 + i * 0.1 });
      await pub.put(payload);
      console.log(`published -> ${k}: ${payload}`);
      if (i + 1 < n) {
        await sleep(interval * 1000);
      }
    }
    await pub.undeclare();
  } finally {
    await s.close();
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: clients/examples/modern/rust/src/bin/publish.rs
//! publish.rs — send data to the goat fabric (modern Rust).
//!
//! ```text
//! export GOAT_ROUTER=tls/127.0.0.1:7447 GOAT_CERT=... GOAT_KEY=... GOAT_CA=... GOAT_NAMESPACE=release/acme
//! cargo run --bin publish              # one JSON sample
//! cargo run --bin publish -- 50 0.2    # 50 samples at 200ms
//! ```
//!
//! Publishes JSON under `<namespace>/sensors/temp`. Real payloads can be anything (bytes,
//! protobuf, CBOR); JSON here for legibility.

use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde_json::json;

#[tokio::main]
async fn main() -> goat_connect::Result<()> {
    zenoh::init_log_from_env_or("error");

    let args: Vec<String> = std::env::args().collect();
    let n: usize = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(1);
    let interval: f64 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(1.0);

    let key = goat_connect::key("sensors/temp")?;

    let session = goat_connect::session().await?;
    let publisher = session.declare_publisher(key.clone()).await?;

    for i in 0..n {
        let ts = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();
        let payload = json!({
            "ts": ts,
            "seq": i,
            "temp_c": 21.5 + (i as f64) * 0.1,
        })
        .to_string();

        publisher.put(payload.clone()).await?;
        println!("published -> {key}: {payload}");

        if i + 1 < n {
            tokio::time::sleep(Duration::from_secs_f64(interval)).await;
        }
    }

    Ok(())
}
